### PR TITLE
Feature/improve social error message

### PR DESCRIPTION
**Pull request by tleekkul**
Return 400 BadRequest when the user is trying to register with existing email.
This will make the client be able to tell the user whether they already login with another login type

### CODE REVIEW

File: ChangeLog [hunk 1]
```diff
@@ -1,3 +1,6 @@
+2018-08-16  Tle Ekkul  <e.aryuth@gmail.com>
+ Returned BadRequest when user's email already exists in the system.
```
Comment (KayEss) [inline ChangeLog]: This is no longer what is returned.

File: Cpp/odin-views/facebook.cpp [hunk 2]
```diff
@@ -73,6 +84,9 @@ namespace {
                 }
                 if ( user_detail.has_key("email") ) {
                     const auto facebook_user_email = fostlib::coerce<fostlib::email_address>(user_detail["email"]);
+                    if ( odin::does_email_exist(cnx, fostlib::coerce<fostlib::string>(user_detail["email"])) ) {
+                        return respond("This email already exists", 400);
```
Comment (KayEss) [inline Cpp/odin-views/facebook.cpp]: Should be 422

File: Cpp/odin-views/google.cpp [hunk 3]
```diff
@@ -52,7 +63,7 @@ namespace {
                     // Use access token as google ID
                     fostlib::insert(user_detail, "sub", access_token);
                     fostlib::insert(user_detail, "name", "Test User");
-                    fostlib::insert(user_detail, "email", access_token + "@mail.com");
+                    fostlib::insert(user_detail, "email", access_token + "@example.com");
```
Comment (KayEss) [inline Cpp/odin-views/google.cpp]: Why are we putting invalid email addresses into the database?

File: Cpp/odin-views/google.cpp [hunk 4]
```diff
@@ -73,6 +84,9 @@ namespace {
                 }
                 if ( user_detail.has_key("email") ) {
                     const auto google_user_email = fostlib::coerce<fostlib::email_address>(user_detail["email"]);
+                    if ( odin::does_email_exist(cnx, fostlib::coerce<fostlib::string>(user_detail["email"])) ) {
+                        return respond("This email already exists", 400);
```
Comment (KayEss) [inline Cpp/odin-views/google.cpp]: 422 not 400

File: Cpp/odin-views/registration.cpp [hunk 5]
```diff
@@ -79,6 +90,9 @@ namespace {
             if ( body.has_key("email") ) {
                 try {
                     const auto email = fostlib::coerce<fostlib::email_address>(body["email"]);
+                    if ( odin::does_email_exist(cnx, fostlib::coerce<fostlib::string>(body["email"])) ) {
+                        return respond("This email already exists", 400);
```
Comment (KayEss) [inline Cpp/odin-views/registration.cpp]: 422 not 400. This repetition should be factored out later on.

File: Cpp/odin/user.cpp [hunk 6]
```diff
@@ -101,3 +101,11 @@ void odin::set_email(
     fostlib::insert(user_values, "email", email.email());
     cnx.insert("odin.identity_email_ledger", user_values);
 }
+
+
+bool odin::does_email_exist(fostlib::pg::connection &cnx, fostlib::string email) {
+    static const fostlib::string sql("SELECT * FROM odin.identity WHERE email=$1");
```
Comment (KayEss) [inline Cpp/odin/user.cpp]: As you're not going through the data, you probably don't want to select and pull all of the column data over the network.

File: Cpp/odin-views/google.cpp (hunk 3 above)
Comment (tleekkul) [reply Cpp/odin-views/google.cpp]: This one is use for mock testing only

Comment (KayEss) [reply Cpp/odin-views/google.cpp]: Oh OK then.

**Pull request closed without merging**